openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
+	}
+
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	*r = Introspection(s.tmp)
+
+	if s.StartedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.StartedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.StartedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.StartedAt = t
+	}
+
+	if s.FinishedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.FinishedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.FinishedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.FinishedAt = t
+	}
+
+	return nil
+}
+
+// GetIntrospectionStatusResult is the response from a Get operation.
+// Call its Extract method to interpret it as an Introspection.
+type GetIntrospectionStatusResult struct {
+	introspectionResult
+}
+
+// StartResult ...
+type StartResult struct {
+	gophercloud.ErrResult
+}
+
+// AbortResult ...
+type AbortResult struct {
+	gophercloud.ErrResult
+}
+
+// Data represents introspection data collected.
+// The format and contents of the stored data depends on the ramdisk used
+// and plugins enabled both in the ramdisk and in inspector itself.
+// This structure has been provided for basic compatibility but it
+// will need extensions
+type Data struct {
+	CPUArch       string                 `json:"cpu_arch"`
+	MACs          []string               `json:"macs"`
+	RootDisk      RootDiskType           `json:"root_disk"`
+	Interfaces    map[string]interface{} `json:"interfaces"`
+	CPUs          int                    `json:"cpus"`
+	BootInterface string                 `json:"boot_interface"`
+	MemoryMB      int                    `json:"memory_mb"`
+	IPMIAddress   string                 `json:"ipmi_address"`
+	Inventory     InventoryType          `json:"inventory"`
+	Error         string                 `json:"error"`
+	LocalGB       int                    `json:"local_gb"`
+	AllInterfaces map[string]interface{} `json:"all_interfaces"`
+}
+
+// RootDiskType ...
+type RootDiskType struct {
+	Rotational         bool        `json:"rotational"`
+	Vendor             string      `json:"vendor"`
+	Name               string      `json:"name"`
+	Hctl               interface{} `json:"hctl"`
+	WwnVendorExtension interface{} `json:"wwn_vendor_extension"`
+	WwnWithExtension   interface{} `json:"wwn_with_extension"`
+	Model              string      `json:"model"`
+	Wwn                interface{} `json:"wwn"`
+	Serial             interface{} `json:"serial"`
+	Size               int         `json:"size"`
+}
+
+// InventoryType ...
+type InventoryType struct {
+	BmcAddress   string                 `json:"bmc_address"`
+	Interfaces   []interface{}          `json:"interfaces"`
+	Disks        []RootDiskType         `json:"disks"`
+	Boot         map[string]interface{} `json:"boot"`
```

Comment:
Similarly for `Boot`: https://github.com/openstack/ironic-python-agent/blob/cfb62c87658e32af373278a8de655367f3b489af/ironic_python_agent/hardware.py#L346

Reply:
A word of caution: do not rely on `map[string]string` for these fields, since we don't guarantee not adding integers or more complex structures later. If we can have a structure with some fields, it may be fine (as long as Go is sane and does not blow up on unexpected fields).

Reply:
Created a BootInfoType for this